fabric/dataclient.go

```diff
@@ -495,170 +413,41 @@ func convertOne(fg *Fabric) ([]*eskip.Route, error) {
-							Args: []interface{}{
-								23, // TODO(sszuecs) needs checking
-							},
-						},
-					},
-					Filters: []*eskip.Filter{
-						{
-							// oauthTokeninfoAnyKV("realm", "/services", "realm", "/employees")
-							Name: filters.OAuthTokeninfoAnyKVName,
-							Args: []interface{}{
-								// TODO(sszuecs): should be configurable
-								"realm",
-								"/services",
-								"realm",
-								"/employees",
-							},
-						},
-						{
-							// oauthTokeninfoAllScope("uid", "spp-brand-service.brands.write")
-							Name: filters.OAuthTokeninfoAllScopeName,
-							Args: privs,
-						},
-						{
-							// unverifiedAuditLog("sub")
-							Name: filters.UnverifiedAuditLogName,
-							Args: []interface{}{
-								// TODO(sszuecs): in the future should be configurable
-								"sub",
-							},
-						},
-					},
-					BackendType: eskipBackend.Type,
-					Backend:     eskipBackend.backend,
-					LBAlgorithm: eskipBackend.lbAlgorithm,
-					LBEndpoints: eskipBackend.lbEndpoints,
+				// local allowlist overrides default. In case allow list is disabled only use scopes.
+				//     example: oauthTokeninfoAnyKV("sub", "app1", "sub", "app2", ..)
+				var allowedServices []interface{}
+				disableAllowList := false
+				if m.AllowList != nil {
+					switch m.AllowList.State {
+					case "disabled":
+						disableAllowList = true
+					default:
+						for _, svcName := range m.AllowList.UIDs {
+							// TODO(sszuecs): in the future "sub" should be configurable
+							allowedServices = append(allowedServices, "sub", svcName)
+						}
+					}
+				} else if len(defaultAllowList) > 0 {
+					allowedServices = append(allowedServices, defaultAllowList...)
 				}
 
-				// add optional ratelimit
-				if m.Ratelimit != nil {
-					r.Filters = append(r.Filters,
-						&eskip.Filter{
-							//inlineContentIfStatus(429, "{\"title\": \"Rate limit exceeded\", \"detail\": \"See the retry-after header for how many seconds to wait before retrying.\", \"status\": 429}", "application/problem+json")
-							Name: filters.InlineContentIfStatusName,
-							Args: []interface{}{
-								429,
-								"{\"title\":\"Rate limit exceeded\",\"detail\":\"See the retry-after header for how many seconds to wait before retrying.\",\"status\":429}",
-								"application/problem+json",
-							},
-						},
-						&eskip.Filter{
-							// clusterClientRatelimit("spp-brand-service_api-brand-assignments-id_DELETE", 30, "1m", "Authorization")
-							Name: filters.ClusterClientRatelimitName,
-							Args: []interface{}{
-								fmt.Sprintf("%s_%s_%s",
-									fg.Metadata.Name,
-									strings.Trim(nonWord.ReplaceAllString(p.Path, "-"), "-"),
-									m.Method,
-								),
-								m.Ratelimit.DefaultRate,
-								m.Ratelimit.Period,
-								// optional header, TODO(sszuecs): maybe configurable in the future
-								"Authorization",
-							},
-						},
-					)
-				}
+				if disableAllowList || len(allowedServices) > 0 || len(defaultAllowList) == 0 {
```

Comment:
complexity comes from https://fabric.docs.zalando.net/fabric-gateway-features/#service-whitelisting (overrides POST , HEAD and PUT) and if no service whitelisting is defined `len(defaultAllowList)=0`

Reply:
Yes. It took good half an hour to go through all the cases and verify correctness by just reasoning with the code.

Reply:
Or alternatively some nice human-oriented godoc/documentation about how it works.

Reply:
I hope I could simplify it better. I could not come up with a condition that does not modify `allowedServices` that is not more complicated than this. For example we can move the `else` into the next `if` but I don't see that this makes anything simpler.
wdyt?

Reply:
I added another commit to refactor the logic into switch case with fallthrough.
I think the most readable solution is the last, wdyt?

Reply:
Followed up with a slightly different proposal for this section. We agreed that it looks more readable.